### PR TITLE
cdc: Remove post-filterings for keys-only/off cdc delta generation

### LINKED ISSUE #7128: cdc: Remove delta_mode::off
CDC information is not useful without at least `delta==keys`, since pre/post image data contains no info on _what_ is actually done in base table operations. 
The addition was a bad (by me - shame!) idea, and we should remove it before we go GA with cdc, so no one uses it by accident.

### PR DESCRIPTION
**Pull request by elcallio**
Refs #7095
Fixes #7128

CDC delta!=full both relied on post-filtering
to remove generated log row and/or cells. This is inefficient.
Instead, simply check if the data should be created in the
visitors.

v2:
    * Fixed delta logs rows created (empty) even when delta == off
v3:
    * Killed delta == off (patch in series)

### CODE REVIEW

File: test/cql/cdc_delta_modes_test.result [hunk 1]
```diff
@@ -24,7 +24,7 @@ select "cdc$batch_seq_no", "cdc$operation", pk, ck, i1 from tb2_scylla_cdc_log w
 			"pk" : "1"
 		},
 		{
-			"cdc$batch_seq_no" : "1",
+			"cdc$batch_seq_no" : "2",
```
Comment (nyh) [inline test/cql/cdc_delta_modes_test.result:27]: Hmm, this change in the test indicates the previous code had a bug, which you fixed now?

**Review by nyh — APPROVED** [review]
Looks good to me, but please have bigger CDC experts than me review it too.

File: test/cql/cdc_delta_modes_test.result (hunk 1 above)
Comment (elcallio) [reply test/cql/cdc_delta_modes_test.result:27]: Nope. That shows that there is a bug in my code. We are inc:ing seq:no one time to many. Woops.

Comment (elcallio) [reply test/cql/cdc_delta_modes_test.result]: fixed.

File: cdc/log.cc [hunk 2]
```diff
@@ -840,6 +840,13 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
     return utils::UUID_gen::get_random_time_UUID_from_micros(t);
 }
 
+/**
+ * Returns whether we should generate cdc delta values (beyond keys)
+ */
+static bool generate_delta_values(const schema& s) {
+    return s.cdc_options().get_delta_mode() == cdc::delta_mode::full;
```
Comment (haaawk) [inline cdc/log.cc:847]: I think this should be decided once when builder is created and not checked every time we want to add something to the result

File: cdc/log.cc [hunk 3]
```diff
@@ -1031,7 +1038,9 @@ struct process_row_visitor {
         bytes value = get_bytes(cell);
 
         // delta
-        _builder.set_value(_log_ck, cdef, value);
+        if (generate_delta_values(_builder.base_schema())) {
```
Comment (haaawk) [inline cdc/log.cc:1041]: I think it would be nicer to have this check hidden inside the builder.

Comment (haaawk) [reply cdc/log.cc:1041]: All those checks.

File: cdc/log.cc [hunk 4]
```diff
@@ -1226,14 +1238,14 @@ struct process_change_visitor {
                 _enable_updating_state, nullptr, &_static_row_state, _clustering_row_states);
         visit_row_cells(v);
 
+        _builder.set_operation(log_ck, operation::update);
```
Comment (haaawk) [inline cdc/log.cc:1241]: What's this and how is this related to this commit?

**Comment by haaawk** [comment]
`cdc: Remove cdc delta_mode::off` looks good to me.

I've got mixed fillings about `cdc: Remove post-filterings for keys-only/off cdc delta generation`. Does it make any difference? Could you generate flamegraphs with and without this commit so that we could see if it's worth doing at all?

File: cdc/log.cc (hunk 4 above)
Comment (elcallio) [reply cdc/log.cc:1241]: rebase error. removing.

File: cdc/log.cc (hunk 2 above)
Comment (haaawk) [reply cdc/log.cc:847]: Or when the visitor is created if we decide not to hide checks inside the builder.

Comment (elcallio) [reply cdc/log.cc:847]: I find it hard to believe that an if per value is gonna cost more than for example even more templating -> code bloat -> icache misses and flushes. 

If a profile really shows that the checks adds measurable cost, we can consider making it embedded (templated), but until then I think that is a premature opt.

**Comment by elcallio** [comment]
@haaawk I assume by flame graph, you mean profile runs. Possibly, but running what? Do we have a standard benchmark for this?
Again, I think you are  _way_ to worried about a few mostly-true branches. Premature opts are the enemy of all things. Except wasting time and code bloat.

File: cdc/log.cc (hunk 2 above)
Comment (haaawk) [reply cdc/log.cc:847]: What I mean is that the visitor (or the builder) should have `bool add_full_deltas = s.cdc_options().get_delta_mode() == cdc::delta_mode::full` set in the constructor and then the checks should be just `if (add_full_deltas) { ...`

**Comment by haaawk** [comment]
> Premature opts are the enemy of all things.

@elcallio exactly my point. I'm saying that `cdc: Remove post-filterings for keys-only/off cdc delta generation` is a premature opts

**Comment by elcallio** [comment]
Yes and no. It has the benefit of simplifying delta==keys, which is good (though to be fair, once delta==off is removed, even the filtrering becomes fairly easy). 
But the filter approach also has the downside of using potentially way more memory than simply not generating the memory in the first place. Which I think has way more performance issues, as it also contaminates caches. (As the saying goes: cache is king!)
How about a compromise: I'll move the check to a local var (delta==off gone makes all code nicer) - that saves two pointer chases at least, and should make the branches virtually costless.

File: cdc/log.cc (hunk 2 above)
Comment (kbr-) [reply cdc/log.cc:847]: `const bool` even! Good for readability.

File: cdc/log.cc [hunk 5]
```diff
@@ -1311,11 +1323,10 @@ struct process_change_visitor {
 
     void partition_delete(const tombstone&) {
         _touched_parts.set<stats::part_type::PARTITION_DELETE>();
-
-        auto log_ck = _builder.allocate_new_log_row(operation::partition_delete);
         if (_enable_updating_state) {
             _clustering_row_states.clear();
         }
+        auto log_ck = _builder.allocate_new_log_row(operation::partition_delete);
```
Comment (kbr-) [inline cdc/log.cc:1329]: why the shuffling around?

Comment (elcallio) [reply cdc/log.cc:1329]: gah, rebase again. fixing.

**Comment by haaawk** [comment]
I'm not worried about the performance of your change (even though a bool variable computed once would be good). I'm more worried about mixing the logic that used to be fully separated with everything else. I don't know how to fix it now. On one hand, pushing the checks down to the builder would hide it from higher abstraction levels but then the builder wouldn't be a single purpose class for building the result any more.

File: cdc/log.cc (hunk 2 above)
Comment (elcallio) [reply cdc/log.cc:847]: Right, ok, agree. Fixing.

**Comment by kbr-** [comment]
Have faith @haaawk. I have some further refactoring plans which might fix your worries.

**Comment by elcallio** [comment]
Moved the check to outermost struct. Now super const member vars.

File: cdc/log.cc [hunk 6]
```diff
@@ -1242,20 +1250,12 @@ struct process_change_visitor {
...
 
-            clustering_row_cells_visitor(
-                    const clustering_key& log_ck, stats::part_type_set& touched_parts, log_mutation_builder& builder,
-                    bool enable_updating_state, const clustering_key* base_ck, cell_map* row_state,
-                    row_states_map& clustering_row_states)
-                : process_row_visitor(
-                        log_ck, touched_parts, builder,
-                        enable_updating_state, base_ck, row_state, clustering_row_states)
-            {}
+            using process_row_visitor::process_row_visitor;
```
Comment (kbr-) [inline cdc/log.cc:1258]: Thanks, I forgot this feature exists.

**Review by kbr- — APPROVED** [review]

**Review by haaawk — APPROVED** [review]
LGTM

**Comment by haaawk** [comment]
Thanks @elcallio

**Comment by nyh** [comment]
@elcallio please rebase, github tells me there are conflicts in the test files

**Comment by elcallio** [comment]
rebased

**Comment by nyh** [comment]
Thanks. Merged to next: f3dfb6e01196c3792e720c4ff6973a4ea53b138d
